File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0 , head)
        prevGroup = dummy 

        while True:
            kth = self.getKth(prevGroup , k)

            if not kth:
                break 

            nextGroup = kth.next 

            #REVERSING THE GROUP

            prev = kth.next 
            cur = prevGroup.next

            while cur != nextGroup:
                temp = cur.next 
                cur.next = prev 
                prev = cur 
                cur = temp

            tmp = prevGroup.next 
            prevGroup.next = kth 
            prevGroup = tmp
        
        return dummy.next


            
        
    #GETTING THE KTH NODE

    def getKth(self , cur , k):
        while cur and k > 0:
            cur = cur.next 
            k -= 1
        return cur 
        

File: 25-reverse-nodes-in-k-group/reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(0 , head);
        ListNode groupPrev = dummy;

        while(true)
        {
            ListNode kth = getKth(groupPrev , k);

            if (kth == null)
            {
                break;
            }

            ListNode groupNext = kth.next;

            //REVERSE THE KTH GROUP 

            ListNode prev = kth.next ;
            ListNode cur = groupPrev.next ;

            while( cur != groupNext)
            {
                ListNode temp = cur.next;
                cur.next = prev;
                prev = cur;
                cur = temp;
            } 

            ListNode temp = groupPrev.next ;
            groupPrev.next = kth ;
            groupPrev = temp ;

        }

        return dummy.next;    
    }

    private ListNode getKth(ListNode cur , int  k)
    {
        while( cur!= null && k > 0 )
        {
            cur = cur.next;
            k -=1;
        }
        return cur;
    }
}